# Exercise 1

In [1]:
!pip install -q pyomo

In [2]:
from pyomo.environ import *

In [3]:
model = ConcreteModel()

In [4]:
model.x1 = Var(domain = Reals)
model.x2 = Var(domain = Reals)
model.x3 = Var(domain = Reals)
model.x4 = Var(domain = Reals)
model.x5 = Var(domain = Reals)
model.x6 = Var(domain = Reals)
model.x7 = Var(domain = Reals)

In [5]:
model.obj = Objective(expr = 2*model.x1 - 5*model.x2 + 14*model.x3 +7*model.x4 - 11*model.x5 + 0.17*model.x6 - 10.2*model.x7, sense = minimize)

In [6]:
model.constr1 = Constraint(expr = model.x1 + model.x2 + model.x3 + model.x4 + model.x5 + model.x6 + model.x7 == 280)
model.constr2 = Constraint(expr = model.x1 - model.x2 >= 10)
model.constr3 = Constraint(expr = -3*model.x3 + model.x6 - 3*model.x7 <= 205)
model.constr4 = Constraint(expr = 20*model.x1 + 42*model.x6 + model.x5 >= 35)
model.constr5 = Constraint(expr = model.x2 - model.x3 + 4*model.x7 >= 24)
model.constr6 = Constraint(expr = model.x1 >= 12)
model.constr7 = Constraint(expr = model.x2 >= 0)
model.constr8 = Constraint(expr = model.x3 >= -5)
model.constr9 = Constraint(expr = model.x4 >= 3)
model.constr10 = Constraint(expr = model.x5 >=10)
model.constr11 = Constraint(expr = model.x6 >= -4)
model.constr12 = Constraint(expr = model.x7 >= 0)

In [7]:
model.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x7 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fi

In [8]:
!apt-get install -y -qq glpk-utils

In [9]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -3035.88
  Upper bound: -3035.88
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 26
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010161161422729492
# ----------------------------------------------------------
#   Solution Information
# ----------------------------

In [10]:
print('Obj =',model.obj())
print('\nDecision Variables :')
print('x1 =',model.x1())
print('x2 =',model.x2())
print('x3 =',model.x3())
print('x4 =',model.x4())
print('x5 =',model.x5())
print('x6 =',model.x6())
print('x7 =',model.x7())
print('\nConstraints :')
print('Constraint 1 :',model.constr1())
print('Constraint 2 :',model.constr2())
print('Constraint 3 :',model.constr3())
print('Constraint 4 :',model.constr4())
print('Constraint 5 :',model.constr5())
print('Constraint 6 :',model.constr6())
print('Constraint 7 :',model.constr7())
print('Constraint 8 :',model.constr8())
print('Constraint 9 :',model.constr9())
print('Constraint 10 :',model.constr10())
print('Constraint 11 :',model.constr11())
print('Constraint 12 :',model.constr12())

Obj = -3035.8799999999997

Decision Variables :
x1 = 12.0
x2 = 0.0
x3 = -5.0
x4 = 3.0
x5 = 269.25
x6 = -4.0
x7 = 4.75

Constraints :
Constraint 1 : 280.0
Constraint 2 : 12.0
Constraint 3 : -3.25
Constraint 4 : 341.25
Constraint 5 : 24.0
Constraint 6 : 12.0
Constraint 7 : 0.0
Constraint 8 : -5.0
Constraint 9 : 3.0
Constraint 10 : 269.25
Constraint 11 : -4.0
Constraint 12 : 4.75


## Ans 3: The value of Objective function is -3035.8799999999997.
The values of the decision variables at the optimal point are as following :


*   $x_1 = 12.0$
*   $x_2 = 0.0$
*   $x_3 = -5.0$
*   $x_4 = 3.0$
*   $x_5 = 269.25$
*   $x_6 = -4.0$
*   $x_7 = 4.75$

##Ans 4:


1.   The value of LHS in Constraint : $\sum_{i=1}^{7}{x_i} = 280$ is 280, hence it is satisfied to equality.
2.   The value of LHS in Constraint : $x_1 - x_2 \geq 10$ is 12. Hence, the constraint is satisfied but not to equality.
3.   The value of LHS in Constraint : $-3x_3 + x_6 - 3x_7 \leq 205$ is -3.25. Hence, the constraint is satisfied but not to equality.
4.   The value of LHS in Constraint : $20x_1 + 42x_6 + x_5 \geq 35$ is 341.25. Hence, the constraint is satisfied but not to equality.











# Exercise 2

##Ans 1: 

For constraint 3 and 4, we assume the denominator to be greater than zero and thus arrive at new constraints :

$-0.8x_1 + 4.8x_2 - 1.8 x_3 + 3.8 x_4 + 2.2 x_5 \leq 0$

$-x_1 + 12x_2 + x_3 + 3x_4 - 5x_5 \leq 0$

In [11]:
exr2_model = ConcreteModel()

In [12]:
exr2_model.x1 = Var(domain = Reals)
exr2_model.x2 = Var(domain = Reals)
exr2_model.x3 = Var(domain = Reals)
exr2_model.x4 = Var(domain = Reals)
exr2_model.x5 = Var(domain = Reals)
exr2_model.y = Var(domain = Reals)

In [13]:
exr2_model.obj = Objective(expr = 0.043*exr2_model.x1 + 0.027*exr2_model.x2 + 0.025*exr2_model.x3 + 0.022*exr2_model.x4 + 0.045*exr2_model.x5 - 0.0275*exr2_model.y, sense = maximize)

In [14]:
exr2_model.constr1 = Constraint(expr = exr2_model.x1 + 3*exr2_model.x2 + exr2_model.x3 + 7*exr2_model.x4 + exr2_model.x5 - exr2_model.y <= 100)
exr2_model.constr2 = Constraint(expr = exr2_model.x2 + 2*exr2_model.x3 + 4*exr2_model.x4 >= 60)
exr2_model.constr3 = Constraint(expr = -0.8*exr2_model.x1 + 4.8*exr2_model.x2 - 1.8*exr2_model.x3 + 3.8*exr2_model.x4 + 2.2*exr2_model.x5 <= 0)
exr2_model.constr4 = Constraint(expr = -exr2_model.x1 + 12*exr2_model.x2 + exr2_model.x3 + 3*exr2_model.x4 - 5*exr2_model.x5 <= 0)
exr2_model.constr5 = Constraint(expr = exr2_model.x1 >= 0)
exr2_model.constr6 = Constraint(expr = exr2_model.x2 >= 0)
exr2_model.constr7 = Constraint(expr = exr2_model.x3 >= 0)
exr2_model.constr8 = Constraint(expr = exr2_model.x4 >= 0)
exr2_model.constr9 = Constraint(expr = exr2_model.x5 >= 0)
exr2_model.constr10 = Constraint(expr = exr2_model.y >= 0)
exr2_model.constr11 = Constraint(expr = exr2_model.y <= 2.25)
exr2_model.constr12 = Constraint(expr = exr2_model.x1 - exr2_model.x2 + exr2_model.x3 - exr2_model.x4 + exr2_model.x5 >= 0)
exr2_model.constr13 = Constraint(expr = exr2_model.x1 + exr2_model.x2 + exr2_model.x3 - exr2_model.x5 >= 0)

In [15]:
exr2_model.pprint()

6 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
  

In [16]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(exr2_model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 3.86940833333333
  Upper bound: 3.86940833333333
  Number of objectives: 1
  Number of constraints: 14
  Number of variables: 7
  Number of nonzeros: 36
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.017251968383789062
# ----------------------------------------------------------
#   Solution Information
# ------------

In [17]:
print('Obj =',exr2_model.obj())
print('\nDecision Variables :')
print('x1 =',exr2_model.x1())
print('x2 =',exr2_model.x2())
print('x3 =',exr2_model.x3())
print('x4 =',exr2_model.x4())
print('x5 =',exr2_model.x5())
print('y =',exr2_model.y())
print('\nConstraints :')
print('Constraint 1 :',exr2_model.constr1())
print('Constraint 2 :',exr2_model.constr2())
print('Constraint 3 :',exr2_model.constr3())
print('Constraint 4 :',exr2_model.constr4())
print('Constraint 5 :',exr2_model.constr5())
print('Constraint 6 :',exr2_model.constr6())
print('Constraint 7 :',exr2_model.constr7())
print('Constraint 8 :',exr2_model.constr8())
print('Constraint 9 :',exr2_model.constr9())
print('Constraint 10 :',exr2_model.constr10())
print('Constraint 11 :',exr2_model.constr11())
print('Constraint 12 :',exr2_model.constr12())
print('Constraint 13 :',exr2_model.constr13())

Obj = 3.869408333333333

Decision Variables :
x1 = 34.9833333333333
x2 = 0.0
x3 = 30.0
x4 = 0.0
x5 = 37.2666666666667
y = 2.25

Constraints :
Constraint 1 : 100.0
Constraint 2 : 60.0
Constraint 3 : 1.1368683772161603e-13
Constraint 4 : -191.3166666666668
Constraint 5 : 34.9833333333333
Constraint 6 : 0.0
Constraint 7 : 30.0
Constraint 8 : 0.0
Constraint 9 : 37.2666666666667
Constraint 10 : 2.25
Constraint 11 : 2.25
Constraint 12 : 102.25
Constraint 13 : 27.71666666666659


## Ans 3 :

The optimal solution value is 3.869408333333333.

The values of all the variables are :


*   $x_1 = 34.983333...$

*   $x_2 = 0.0$

*   $x_3 = 30.0$
*   $x_4 = 0.0$


*   $x_5 = 37.26666666667$


*   $y = 2.25$

Activity of constraints :


1.   The value of LHS in Constraint : $x_1 + 3x_2 + x_3 + 7x_4 + x_5 - y \leq 100$ is 100. Hence, the constraint is satisfied to equality.
2.   The value of LHS in Constraint : $x_2 + 2x_3 + 4x_4\geq 60$ is 60. Hence, the constraint is satisfied to equality.
3.   The value of LHS in Constraint : $-0.8x_1 + 4.8x_2 - 1.8 x_3 + 3.8 x_4 + 2.2 x_5 \leq 0$ is 1.1368683772161603e-13. Hence, the constraint is satisfied but not to equality.
4.   The value of LHS in Constraint : $-x_1 + 12x_2 + x_3 + 3x_4 - 5x_5 \leq 0$ is -191.3166666666668. Hence, the constraint is satisfied but not to equality.
5.   The value of LHS in Constraint : $x_1 \geq 0$ is 34.9833333333333. Hence, the constraint is satisfied but not to equality.
6.   The value of LHS in Constraint : $x_2 \geq 0$ is 0. Hence, the constraint is satisfied to equality.
7.   The value of LHS in Constraint : $x_3 \geq 0$ is 30.0. Hence, the constraint is satisfied but not to equality.
8.   The value of LHS in Constraint : $x_4 \geq 0$ is 0.0. Hence, the constraint is satisfied to equality.
9.   The value of LHS in Constraint : $x_5 \geq 0$ is 37.2666666666667. Hence, the constraint is satisfied but not to equality.
10.   The value of LHS in Constraint : $y \geq 0$ is 2.25. Hence, the constraint is satisfied but not to equality.
11.   The value of LHS in Constraint : $y \leq 0$ is 2.25. Hence, the constraint is satisfied to equality.
12.   The value of LHS in Constraint : $x_1 - x_2 + x_3 - x_4 + x_5 \geq 0$ is 102.25. Hence, the constraint is satisfied but not to equality.
13.   The value of LHS in Constraint : $x_1 + x_2 + x_3 - x_5 \geq 0$ is 27.71666666666659. Hence, the constraint is satisfied but not to equality.




## Ans 4 :

In case we do not make any assumptions, we will solve the model for all the four cases relating to the denominators in constraint 3 and 4 and choose the solution accordingly.
So, in this case we will get 4 linear programs corresponding to each case relating to the denominators in constraint 3 and 4.

# Exercise 3

## Ans 1 : Formulation as a linear program:

$\text{Decision Variables : p = production of product P in this month.}\\\text{q = production of product Q in this month.}$

$\text{Objective Function : min 145p + 280q}\\
\text{subject to: p} \geq {84}\\
\text{q}\geq80\\
\text{p + q}\geq{274}\\
\text{69p + 55q}\leq{25800}\text{     (changed the time to minutes for easier calculation)}$



In [18]:
exr3_model = ConcreteModel()

In [19]:
exr3_model.prod_p = Var(domain = Reals)
exr3_model.prod_q = Var(domain = Reals)

In [20]:
exr3_model.cost = Objective(expr = 145*exr3_model.prod_p + 280*exr3_model.prod_q, sense = minimize)

In [21]:
exr3_model.prod_minimum = Constraint(expr = exr3_model.prod_p + exr3_model.prod_q >= 110 + 84 +80)
exr3_model.prod_p_minimum = Constraint(expr = exr3_model.prod_p >= 84)
exr3_model.prod_q_minimum = Constraint(expr = exr3_model.prod_q >= 80)
exr3_model.time_maximum = Constraint(expr = 69 * exr3_model.prod_p + 55 * exr3_model.prod_q <= 25800)

In [22]:
exr3_model.pprint()

2 Var Declarations
    prod_p : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    prod_q : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 145*prod_p + 280*prod_q

4 Constraint Declarations
    prod_minimum : Size=1, Index=None, Active=True
        Key  : Lower : Body            : Upper : Active
        None : 274.0 : prod_p + prod_q :  +Inf :   True
    prod_p_minimum : Size=1, Index=None, Active=True
        Key  : Lower : Body   : Upper : Active
        None :  84.0 : prod_p :  +Inf :   True
    prod_q_minimum : Size=1, Index=None, Active=True
        Key  : Lower : Body   : Upper : Active
        None :  80.0 : prod_q :  +Inf :   True
    time_ma

In [23]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(exr3_model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 50530.0
  Upper bound: 50530.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 3
  Number of nonzeros: 7
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.012539386749267578
# ----------------------------------------------------------
#   Solution Information
# --------------------------------

In [24]:
print('Cost = ',exr3_model.cost())

print('\nDecision Variables')
print('Production of P = ',exr3_model.prod_p())
print('Production of Q = ',exr3_model.prod_q())

print('\nConstraints')
print('Total Production  = ',exr3_model.prod_minimum())
print('Production of P  = ',exr3_model.prod_p_minimum())
print('Production of Q  = ',exr3_model.prod_q_minimum())
print('Time used for Production (in hrs)  = ',exr3_model.time_maximum()/60)

Cost =  50530.0

Decision Variables
Production of P =  194.0
Production of Q =  80.0

Constraints
Total Production  =  274.0
Production of P  =  194.0
Production of Q  =  80.0
Time used for Production (in hrs)  =  296.43333333333334


## Ans 3 : 
The optimal value of Cost is 50530.

The values of the variables i.e. production of each product in September:


1.   p = 194
2.   q = 80

Activity of constraints :


1.   The value of LHS in constraint 1 i.e. total production of cost ($\text{p + q}\geq{274}$) is 274. Hence, it satisfies the constraint to equality.
2.   The value of LHS in constraint 2 i.e. minimum production of p ($\text{p} \geq {84}$) is 194, hence it satisfies the constraint but not to equality.
3.   The value of LHS in constraint 3 i.e. minimum production of q ($\text{q} \geq {80}$) is 80, hence it satisfies the constraint to equality.
4.   The value of LHS in constraint 4 i.e. maximum time required for production of p and q ($\text{69p + 55q}\leq{25800}\text{(430 hrs)}$) is 296.43333333333334 (in hrs), hence it satisfies the constraint but not to equality.



## Ans 4 :

After the industry recieves another order, the linear program would look like :

$\text{Decision Variables : p = production of product P in this month.}\\\text{q = production of product Q in this month.}$

$\text{Objective Function : min 145p + 280q}\\
\text{subject to: p} \geq {186}\text{        (We add the new order into the previous order)}\\ 
\text{q}\geq{197}\text{           (We add the new order into the previous order)}\\
\text{p + q}\geq{493}\text{           (We add the new order into the previous combined minimum production)}\\
\text{69p + 55q}\leq{25800}\text{     (changed the time to minutes for easier calculation)}$ 

In [25]:
exr3_model.prod_p_minimum_revised = Constraint(expr = exr3_model.prod_p >= 84 + 102)
exr3_model.prod_q_minimum_revised = Constraint(expr = exr3_model.prod_q >= 80 + 117)
exr3_model.prod_minimum_revised = Constraint(expr = exr3_model.prod_p + exr3_model.prod_q >= 110 + 84 + 80 + 102 + 117)

In [26]:
exr3_model.pprint()

2 Var Declarations
    prod_p : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None : 194.0 :  None : False : False :  Reals
    prod_q : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  80.0 :  None : False : False :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 145*prod_p + 280*prod_q

7 Constraint Declarations
    prod_minimum : Size=1, Index=None, Active=True
        Key  : Lower : Body            : Upper : Active
        None : 274.0 : prod_p + prod_q :  +Inf :   True
    prod_minimum_revised : Size=1, Index=None, Active=True
        Key  : Lower : Body            : Upper : Active
        None : 493.0 : prod_p + prod_q :  +Inf :   True
    prod_p_minimum : Size=1, Index=None, Active=True
        Key  : Lower : Body   : Upper : Active
        None :  84.0 : prod_p :  +

In [27]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(exr3_model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 8
  Number of variables: 3
  Number of nonzeros: 11
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.014358758926391602


## Ans 6 : 

We don't attain an optimal solution after the introduction of the new order.

This happens because the time required the produce the total production after the new order requires more time than what is available.

Based on the results obtained, Maple Industries should either increase the manufacturing time or decline the new order.
